### PR TITLE
Add manifest for building as a flatpak

### PR DESCRIPTION
**Pull request by cpba**
Includes the following:

- `org.DolphinEmu.dolphin-emu.metainfo.xml.in` this file contains general information and metadata that is used by most linux based app stores (kde discover, gnome software, flathub.org, etc) to showcase dolphin.

- `fill_release_node.sh` a script to automatically fill in the release information in the metainfo file when building the flatpak.
- `SDL2.json` manifest to build the vendored version of SDL from the `Exports` directory, it's only temporarily needed until the kde runtime is updated with the current SDL2 version and should be deleted after.
- `org.DolphinEmu.dolphin-emu.yml` the manifest itself.

### CODE REVIEW

**Comment by cpba** [comment]
Some other things I'm not sure about:

Regarding the metainfo file, it contains a `releases` node that can be used to hold some data on the current release. On flathub this node was automatically updated by the flathub-checker bot using data read from https://dolphin-emu.org/update/latest/beta but that wouldn't make sense here.

Perhaps the best solution would be to place it in the data directory and fill the release node during the cmake build. The file itself is an xdg standard and not flatpak specific so it would also be useful for deb/rpm and other kinds of builds.

Would this be a good solution?

The file also includes links to some screenshots, this are the screenshots that are shown on flathub and other app stores: https://flathub.org/apps/org.DolphinEmu.dolphin-emu
These screenshots are currenly hosted on flathub, but it would make more sense to move them somewhere on dolphin-emu.org. They are also quite old and low res, so it might be good to replace them with newer ones.

Are there any that you'd like to use?

**Comment by ColinKinloch** [comment]
`SDL2-no-libdecor` shouldn't be needed any more.
There are a few updates since `23.08` but I'm not sure any are relevant to dolphin.

**Comment by cpba** [comment]
> `SDL2-no-libdecor` shouldn't be needed any more. There are a few updates since `23.08` but I'm not sure any are relevant to dolphin.

It's not necessary for dolphin-emu `2407`, but versions `2407-232` and newer depend on [2.30.6](https://github.com/khg8m3r/dolphin/blob/d2c4f3af1958c26e50734cb04bd9442cf72eb5ef/Externals/SDL/CMakeLists.txt#L12C12-L12C13), which made me bundle it because `org.kde.Platform` was only shipping 2.28.5 which I assumed was the same version as fdo `23.08`.

Do you know if it has been updated recently?

**Comment by ColinKinloch** [comment]
Ah, never mind then.
I wonder what aspect of `2.30.6` specifically dolphin depends on.
It looks like freedesktop `24.08beta` is shipping with `2.30` so it may only be a matter of days.

**Comment by cpba** [comment]
> I wonder what aspect of `2.30.6` specifically dolphin depends on.

This is all I could find: https://github.com/dolphin-emu/dolphin/pull/12144#issuecomment-2267201248

Most likely talking about the crash when a controller is disconnected mentioned in the [release notes](https://github.com/libsdl-org/SDL/releases/tag/release-2.30.6).

> It looks like freedesktop 24.08beta is shipping with 2.30 so it may only be a matter of days.

Do you know if there'll be a `6.7-24.08` runtime?

**Comment by ColinKinloch** [comment]
Looking at the output of `flatpak install org.kde.Platform`: the `5.15` is the version that gets a suffix.
Which I guess implies that `24.08` will be the base for `6.8` or `6.7` will get updated.
The answer is probably here somewhere: https://invent.kde.org/packaging/flatpak-kde-runtime

**Comment by ColinKinloch** [comment]
I'd guess its more a matter of a couple of months rather than days 😅

**Review by OatmealDome — COMMENTED** [review]
Sorry about the delay. I've been occupied by other projects.

Also, extremely minor nitpick: I think we should name the containing folder `Flatpak` to match the capitalization of the other folders.

File: flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml
Comment (cpba) [reply flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml:2]: Jeremy Newton is the maintainer of the fedora package from where this file was originally copied back in 2017, It's very likely that it has been entirely rewritten since then.

Changing the copyright is unlikely to be an issue if need be.

File: flatpak/org.DolphinEmu.dolphin-emu.yml
Comment (cpba) [reply flatpak/org.DolphinEmu.dolphin-emu.yml:78]: You're right, since this build is official it makes no sense to list `Flathub` as the distributor.
Since flathub doesn't require thism I'll just change it to `dolphin-emu.org`.

File: flatpak/org.DolphinEmu.dolphin-emu.yml
Comment (cpba) [reply flatpak/org.DolphinEmu.dolphin-emu.yml:26]: This is frowned upon on flathub because it requires doing a full clone with all the submodules for each build, which lengthens build times and eats into shared resources.

But since this'll run on ci with full sources that's not a concern, I'll change the manifest to use the local directory rather than cloning the latest git commit and use the vendored version.

File: flatpak/org.DolphinEmu.dolphin-emu.yml
Comment (cpba) [reply flatpak/org.DolphinEmu.dolphin-emu.yml:69]: Yes, same as with libusb.

Although in this case building sdl2 will only be a stopgap measure of a few weeks until the kde runtime catches up, I'll change the separate file to use the vendored version rather than downloading the tarball.

Once the runtime gets 2.30.6 we can just delete the file.

File: flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml
Comment (cpba) [reply flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml:91]: I can't believe that email is still in there ;)

Yes, of course, it can be changed to anything you want.

Which address should it be?

File: flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml
Comment (cpba) [reply flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml]: Alright, I'll just drop it!

File: flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml
Comment (cpba) [reply flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml]: Fixed!

File: flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml
Comment (cpba) [reply flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml]: > We'll definitely need to rehost these screenshots on a Dolphin controlled server.

Would the best option be somewhere on the dolphin website? Should I open another pr there?

>  I'll ask around and see if we can take some newer ones.

Thank you!

Although, if you don't find any I could just take new ones.

Comment (OatmealDome) [reply flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml]: > Would the best option be somewhere on the dolphin website? Should I open another pr there?

No need! I've uploaded the screenshots here:

```
http://dolphin-emu.org/m/user/flatpak/screenshot_1.png
http://dolphin-emu.org/m/user/flatpak/screenshot_2.png
http://dolphin-emu.org/m/user/flatpak/screenshot_3.png
http://dolphin-emu.org/m/user/flatpak/screenshot_4.png
```

I think we can leave replacing these screenshots to a later time. I wouldn't want to bikeshed this PR any further.

File: flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml
Comment (OatmealDome) [reply flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml]: Does this e-mail show up in the KDE Discover / flathub.org / etc UI anywhere?

File: Flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml
Comment (OatmealDome) [inline Flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml:39]: I think we can par this down to just 2407 and 2409?

File: flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml
Comment (cpba) [reply flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml]: Not that I know of.

It's optional and of dubious utility so I'll just drop it.

File: Flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml
Comment (cpba) [reply Flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml:39]: Flathub only displays the top one anyway, we can reduce it to a single one.

It's only like this because the flathub bot that submits the new versions doesn't delete the older nodes.

I'm a bit more worried about updating it like I was saying here https://github.com/dolphin-emu/dolphin/pull/13040#issuecomment-2316138250 I'm not sure how to best update it.

Should I keep it flatpak-only for now?

File: flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml
Comment (cpba) [reply flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml]: Alright, done!

File: Flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml
Comment (OatmealDome) [reply Flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml]: Oops, I missed that comment.

I agree that it would be nice to automatically populate this field at build time. [I can see that PCSX2 uses some placeholders here that are replaced by a script at build time](https://github.com/PCSX2/pcsx2/blob/master/.github/workflows/scripts/linux/pcsx2-qt.metainfo.xml.in). Perhaps we can use the same system here too?

Comment (cpba) [reply Flatpak/org.DolphinEmu.dolphin-emu.metainfo.xml]: Great, we'll keep it flatpak-only for now.

We can always move it to generic linux builds in the future if necessary.

**Comment by OatmealDome** [comment]
Could you drop the final commit and squash everything?

About the release node commit: I believe you copied the `generate-metainfo.sh` script near verbatim from PCSX2. Unfortunately, that codebase is licensed under GPLv3, while Dolphin's code is GPLv2+. This might pose a licensing problem. I don't want to hold up this PR any further, so I think we should leave that out for now and fix it in a later PR.

**Comment by cpba** [comment]
Sorry I made a mess of the rebase, my tracking branch was on the wrong commit. I'll fix it.

**Comment by cpba** [comment]
> Could you drop the final commit and squash everything?
> 
> About the release node commit: I believe you copied the `generate-metainfo.sh` script near verbatim from PCSX2. Unfortunately, that codebase is licensed under GPLv3, while Dolphin's code is GPLv2+. This might pose a licensing problem. I don't want to hold up this PR any further, so I think we should leave that out for now and fix it in a later PR.

Apologies I didn't realise the licensing would be incompatible.

It's a rather trivial shell script so I've written a simple four liner to replace it, I hope that's fine, if it isn't I can just delete it and squash again.

Thank you for your help, what's the next step? Should I be looking into integrating it with the ci?

**Comment by OatmealDome** [comment]
I'll handle the CI infrastructure, so don't worry about that. Thanks for your work on the PR. Let me ask around and see if anyone has any problems with this. If there aren't, I'll merge.

**Comment by cpba** [comment]
I fixed two small issues:

- `libevdev`  was one point release behind the one currently being used on flathub.
- `fill_release_node.sh` would produce a version string that included the short hash of the commit.

I haven't squashed again in case you'd rather drop them.

**Comment by OatmealDome** [comment]
Sorry about the delay! Just wanted to give you a heads up that I plan to merge this within the next day or so now, now that all of our CI infrastructure is ready.

**Comment by dreamsyntax** [comment]
> SDL2.json manifest to build the vendored version of SDL from the Exports directory, it's only temporarily needed until the kde runtime is updated with the current SDL2 version and should be deleted after.

Since its been a few months, is this still needed?

**Comment by ColinKinloch** [comment]
> Since its been a few months, is this still needed?

Yes.

However it's now a [matter of days](https://invent.kde.org/packaging/flatpak-kde-runtime/-/merge_requests/228#note_1057480)?

**Review by dreamsyntax — APPROVED** [review]

**Comment by cpba** [comment]
We [haven't been able to update the runtime to 6.8](https://github.com/flathub/org.DolphinEmu.dolphin-emu/pull/288) because it's not available yet.

I'll open a pr to drop the vendored sdl and upgrade the runtime as soon as it is.

**Comment by OatmealDome** [comment]
This is now live!

* Dev builds are available at this repo: `https://flatpak.dolphin-emu.org/dev.flatpakrepo`
* PR builds are available at this repo: `https://flatpak.dolphin-emu.org/prs.flatpakrepo` (each PR is its own branch named `pr-xxxxx`)
* Single-file bundles are also available at `dolphin-emu.org`

**Pull request merged**
